design/crypto-engines.md

```diff
@@ -0,0 +1,148 @@
+# Crypto Engines
+
+The Go standard library crypto library is very well done and widely used.
+The problem, however, comes when a user wants to swap out the implementation
+for another. This has already happened even upstream with the dev.boringcrypto [0]
+branch. Use cases for using an alternate implementation include FIPS compliance
+(where a user may need to call into a FIPS certified crypto library), and
+interfacing with external crypto devices (e.g. accelerator cards, specialty
+hardware).
+
+It would be easy enough to maintain a package that does this already, however
+the problem comes when trying to swap crypto implementation in a project
+you do not own when building against 3rd party code. This would also ensure
+consistency so that there is a formal way to declare that all or part of the
+crypto operations provided by the standard library should be implemented
+by some outside package.
+
+This proposal outlines a way for alternate implementations of the
+crypto functionality to be registered via a function call, and allow the
+alternate crypto implementation to live in a separate package maintained
+outside of the Go tree.
+
+## Engine implementation
+
+Users should be able to register a single global engine which if present
+will be used instead of the standard library implementation. The design allows
+engines to replace all *or* part of the standard library implementations.
+
+The loading and registering of an engine should be simple and straighforward.
+The user is presented with 2 options:
+
+1. Load an engine dynamically by having the engine source code and dynamically
+   registering the engine struct itself, or
+2. Load an engine via Go plugin functionality [1]
+
+The basic outline for these functions would look like:
+
+```
+package crypto
+
+var engine Engine
+
+// RegisterEngine registers the engine specified by `e`.
+// This engine will be used for any crypto functionality
+// it provides, falling back to the standard library for
+// any functionality not provided by the engine.
+func RegisterEngine(e Engine) {
+	if engine != nil {
+		panic("engine already registered")
+	}
+	engine = e
+}
+
+// RegisterEnginePlugin will load the plugin located
+// at `path` and subsequently create and register 
+// an engine from it.
+func RegisterEnginePlugin(path string) {
+	...
+}
+```
+
+It would also be very easy to unregister an Engine, if need be, via:
+
+```
+func EngineCleanup() {
```

Comment:
what happen if  goroutine calls this cleanup function while another has handles held ?
Say because the other it is using a stream cipher, or a TLS channel is currently open ?